python/graphstorm/model/node_glem.py

```diff
@@ -0,0 +1,267 @@
+            params = self.lm.parameters()
+        elif part == 'gnn':
+            params = self.gnn.parameters()
+        for param in params:
+            param.requires_grad = False
+
+    def unfreeze_params(self, part='lm'):
+        """Unfreeze parameters in lm or gnn"""
+        if part == 'lm':
+            params = self.lm.parameters()
+        elif part == 'gnn':
+            params = self.gnn.parameters()
+        for param in params:
+            param.requires_grad = True
+
+    def toggle(self, part='lm'):
+        """The method toggles training between lm and gnn."""
+        if part == 'lm':
+            self.training_lm = True
+            self.freeze_params('gnn')
+            self.unfreeze_params('lm')
+        elif part == 'gnn':
+            self.training_lm = False
+            self.freeze_params('lm')
+            self.unfreeze_params('gnn')
+
+    def forward(self, blocks, node_feats, edge_feats, labels, input_nodes, use_gnn=True):
+        """ Forward pass for GLEM model.
+        Parameters
+        ----------        
+        blocks : List[dgl.heterograph.DGLBlock]
+        node_feats : Dict[ntype: tensors.shape [bs, feat_dim]]
+        edge_feats : None
+        labels : Dict[target_ntype: tensor.shape [bs]]
+        input_nodes : {target_ntype: tensor.shape [bs], other_ntype: []}
+        use_gnn : bool
+            If True, use GNN's decoder, otherwise, use LM's decoder 
+        """
+        if use_gnn:
+            total_loss = self.forward_gnn(blocks, node_feats, edge_feats, labels, input_nodes)
+        else:
+            total_loss = self.forward_lm(blocks, node_feats, edge_feats, labels, input_nodes)
+        return total_loss
+
+    def forward_gnn(self, blocks, node_feats, _, labels, input_nodes):
+        """Forward pass for node prediction using GNN
+        """
+        emb = self._embed_nodes(blocks, node_feats, _, input_nodes)
+        labels = self._process_labels(labels)
+        logits = self.gnn.decoder(emb)
+
+        batch_size = labels.size(0)
+        n_gold_nodes = batch_size // 2
+        # compute pseudo labels from LM
+        logits_lm = self.lm.decoder(emb[n_gold_nodes:])
+        pseudo_labels = logits_lm.argmax(-1)
+
+        # GLEM loss
+        loss = compute_loss(self.gnn.loss_func, logits, labels[:n_gold_nodes], pseudo_labels,
+                            pl_weight=self.pl_weight)
+
+        # add regularization loss to all parameters to avoid the unused parameter errors
+        reg_loss = th.tensor(0.).to(loss.device)
+        # L2 regularization of dense parameters
+        for d_para in self.gnn.get_dense_params():
+            reg_loss += d_para.square().sum()
+        # weighted addition to the total loss
+        return loss + self.alpha_l2norm * reg_loss
+
+
+    def forward_lm(self, blocks, node_feats, _, labels, input_nodes):
+        """Forward pass for node prediction using LM"""
+        emb = self._embed_nodes(blocks, node_feats, _, input_nodes)
```

Comment:
In GLEM, we also need to compute GNN for both E and M stages? if so, it'll be very expensive to fine-tune the BERT model

Reply:
Yes. In E stage (LM fine-tuning), need to run GNN to generate pseudo labels for supervising LM.